### PR TITLE
fix: handle base href

### PR DESCRIPTION
**Pull request by aarongranick-okta**
- removes dependency on `configuration-validation` as it was causing warnings about commonjs modules
- better handling of `originalUri` and base href

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -46,9 +46,14 @@ export default {
             browser: true
         }),
         commonjs({
-            // namedExports: {
-            //     'node_modules/@okta/okta-auth-js/dist/okta-auth-js.umd.js': ['toRelativeUrl']
-            // }
+            namedExports: {
+                // TODO: why is this necessary?
```
Comment (shuowu) [inline rollup.config.js:50]: https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module

**Comment by shuowu** [comment]
I originally thought we can let users pass in the `restoreOriginalUri` callback to handle different use cases (like base href, other routing strategy). What the users need to do is to use the `originalUri` from the input, then massage it with whatever base or routing strategy. So we don't need to handle the complex URL handling logic.
https://github.com/okta/okta-angular/blob/66de5c492e8f90417d1122a2e16d82d4ce61e660/src/okta/services/okta.service.ts#L58-L62

**Comment by aarongranick-okta** [comment]
> I originally thought we can let users pass in the `restoreOriginalUri` callback to handle different use cases (like base href, other routing strategy). What the users need to do is to use the `originalUri` from the input, then massage it with whatever base or routing strategy. So we don't need to handle the complex URL handling logic.

I thought it was pretty straight-forward as well, but spending more time with samples I discovered a few issues that make it not so simple:

- In angular all the urls should be relative paths, and match router definitions. If they don't, an error will happen. 
- Angular can be started with a `--base-href` option. This will automatically add a prefix for router. This prefix can be discovered by calling `location.prepareExternalUrl`. The `Location` service is available via angular injection - HOWEVER -
- Okta config and route definitions are defined BEFORE any Angular injection is available. So neither Okta config nor Angular routes can be a function of the `Location` service.

In order to work correctly with a base href, you will need to have a value for both `redirectUri` and `postLogoutRedirectUri` that have this base href included. These are part of Okta config, so this is a problem.

A basic login button will call `signInWithRedirect` which will set `originalUri` - by default - to `window.location.origin`. This is also a problem. The user would have to ALWAYS call this method with an `originalUri` parameter, the default will not work, or will break in strange ways.

So the solution here has a few parts:

- restoreOriginalUri will convert absolute to relative before calling `router.navigateToUrl` - this is important because an absolute URL will break. However, to know the true baseUrl, we also need to take the base href into account. Calling the router with the base href included will also break. It needs to be a path that is defined in the routes definition
- signInWithRedirect will set a default originalUri of `/`. This is equivalent in meaning to `window.location.origin` - we want to navigate back to the **app root**, which may or may not have a base href.
- signOut will set a default `postLogoutRedirectUri` which is an absolute URL to the app root, containing base href if set.
- the constructor will convert a relative `redirectUri` into an absolute URL, including base href.

With these changes, support for `--base-href` is very straightforward. All you need to do is use a relative url `/login/callback` for your `redirectUri`, all the other details should work right out of the box.

**Comment by shuowu** [comment]
now I realize `--base-href` is an `ng` options ... 
LGTM!

**Review by shuowu — APPROVED** [review]

**Review by shuowu-okta — APPROVED** [review]

File: rollup.config.js (hunk 1 above)
Comment (swiftone) [reply rollup.config.js:50]: Can we update this comment with an answer?

**Review by swiftone — APPROVED** [review]

File: rollup.config.js (hunk 1 above)
Comment (aarongranick-okta) [reply rollup.config.js:50]: This link explains why the error could be shown. I still don't understand why it is necessary. `okta-auth-js` is not a commonjs module. and these namedExports only seem to affect functions from the util file, other named exports are handled fine. So there is still a mystery here. When we figure it out, we should be able to remove this or explain it better.

**Pull request closed without merging**